basic_pipeline/2_SystemArchitecture.md

```diff
@@ -0,0 +1,25 @@
+Once we know what should be done, let's start thinking how should our system look like!
+
+First, let's agree on the naming convention. In the Membrane Framework, there is a concept of the `pipeline` which consists of multiple `elements`. Elements are linked with the use of the `pads`, which can be input pads or output pads.
+Depending on what type of pads the particular element is equipped with, we distinguish the following types of elements:
++ Source - element with only output pads, the first element of each pipeline. It is responsible for fetching the data and transmitting it through the output pad.
++ Filter - element with both the input pads and the output pads. Its purpose is to get the data via the input pad, transform it and finally transmit the processed data via the output pad.
++ Sink - element with only input pads, the last element of the pipeline. It can be responsible for storing or playing the data received via the input pad. 
+Between the pads, the `buffers` are being sent. 
+Buffers sent between the pads should be of the same type. In the Membrane Framework nomenclature, we say, that the pads should have corresponding capabilities - `caps`.
+One might wonder when does the event of the buffers being sent occurs - well, it depends on the pad's mode!
+If the pad works in the *pull* mode, the buffer is sent when it is demanded by the succeeding element in the pipeline.
+Otherwise, the pad works in the *push* mode, which means that it is pushing the buffers once they are ready, independently from the fact if the following elements want (and are capable of processing) it.
+In our pipeline, all the pads will work in the *pull* mode, which inducts a flow control by a backpressure mechanism.
+# Scheme
+![Pipeline scheme](assets/images/basic_pipeline.png) <br>
+As you can see, our pipeline will consist of two twin branches, one per each of the peers. The branches will be merged with the `Mixer` element and the result produced by this element will be put in the file with the `Sink` elements. 
+Here you can find the description of the particular elements of the system.
+# Elements description
++ **Source** - that element is responsible for reading the list of packets from the file. 
++ **Ordering Buffer** - this element is responsible for reordering the packets based on their sequence id. The most important part of this element is the buffer, within which there is a place for all the packets, identified by the packet's sequence id. Once the new packet is received, it is placed in the proper place in the buffer, depending on the sequence id. If there is a consistent part of packets "a the bottom of the buffer" (which means - packets with the subsequent sequence ids, which hasn't already been sent yet), then this part is sent via the output pad. That is how we make sure that the next element will receive elements sorted by the sequence id. 
```

Comment:
typo: "a the bottom ..." -> "at the bottom"

Reply:
indeed, thanks for spotting!
Just for the future - you can propose a commit straight on the github page - for such typos it might be more convenient for both of us, as you could simply fix a typo instead of pointing where the typo is, and I could just accept such a commit :D

Reply:
done